apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -249,7 +250,12 @@ handle_stream_start({xmlstreamstart, _Name, Attrs}, #state{} = S0) ->
         {?NS_STREAM, true} ->
             change_shaper(S, jid:make(<<>>, Server, <<>>)),
             Version = xml:get_attr_s(<<"version">>, Attrs),
-            stream_start_by_protocol_version(Version, S);
+            SpamCtlState = ejabberd_hooks:run_fold(spamctl_initialise,
```

Comment:
These `spamctl_*` set of hooks are quite powerful. How about passing whole c2s state as an arg to the `initialise` hook handler so the implementation may choose what to do with it? Other solution would be passing the state to all the hooks. I can imagine that some clever logic (rewording or stopping the user) may need some of the fields which are in c2s state.